Mathlib/FieldTheory/Relrank.lean

```diff
@@ -0,0 +1,527 @@
+
+theorem relfinrank_eq_toNat_relrank : relfinrank A B = toNat (relrank A B) := rfl
+
+variable {A B C}
+
+theorem relrank_eq_of_inf_eq (h : A ⊓ C = B ⊓ C) : relrank A C = relrank B C := by
+  simp_rw [relrank]
+  congr!
+
+theorem relfinrank_eq_of_inf_eq (h : A ⊓ C = B ⊓ C) : relfinrank A C = relfinrank B C :=
+  congr(toNat $(relrank_eq_of_inf_eq h))
+
+/-- If `A ≤ B`, then `Subfield.relrank A B` is `[B : A]` -/
+theorem relrank_eq_rank_of_le (h : A ≤ B) : relrank A B = Module.rank A (extendScalars h) := by
+  rw [relrank]
+  have := inf_of_le_left h
+  congr!
+
+/-- If `A ≤ B`, then `Subfield.relfinrank A B` is `[B : A]` -/
+theorem relfinrank_eq_finrank_of_le (h : A ≤ B) : relfinrank A B = finrank A (extendScalars h) :=
+  congr(toNat $(relrank_eq_rank_of_le h))
+
+variable (A B C)
+
+theorem inf_relrank_right : relrank (A ⊓ B) B = relrank A B :=
+  relrank_eq_rank_of_le (inf_le_right : A ⊓ B ≤ B)
+
+theorem inf_relfinrank_right : relfinrank (A ⊓ B) B = relfinrank A B :=
+  congr(toNat $(inf_relrank_right A B))
+
+theorem inf_relrank_left : relrank (A ⊓ B) A = relrank B A := by
+  rw [inf_comm, inf_relrank_right]
+
+theorem inf_relfinrank_left : relfinrank (A ⊓ B) A = relfinrank B A :=
+  congr(toNat $(inf_relrank_left A B))
+
+@[simp]
+theorem relrank_self : relrank A A = 1 := by
+  rw [relrank_eq_rank_of_le (le_refl A), extendScalars_self, IntermediateField.rank_bot]
+
+@[simp]
+theorem relfinrank_self : relfinrank A A = 1 := by
+  simp [relfinrank_eq_toNat_relrank]
+
+variable {A B} in
+theorem relrank_eq_one_of_le (h : B ≤ A) : relrank A B = 1 := by
+  rw [← inf_relrank_right, inf_eq_right.2 h, relrank_self]
+
+variable {A B} in
+theorem relfinrank_eq_one_of_le (h : B ≤ A) : relfinrank A B = 1 := by
+  simp [relfinrank_eq_toNat_relrank, relrank_eq_one_of_le h]
+
+variable {A B} in
+theorem relrank_mul_rank_top (h : A ≤ B) : relrank A B * Module.rank B E = Module.rank A E := by
+  rw [relrank_eq_rank_of_le h]
+  letI : Algebra A B := (inclusion h).toAlgebra
+  haveI : IsScalarTower A B E := IsScalarTower.of_algebraMap_eq' rfl
+  exact rank_mul_rank A B E
+
+variable {A B} in
+theorem relfinrank_mul_finrank_top (h : A ≤ B) : relfinrank A B * finrank B E = finrank A E := by
+  simpa using congr(toNat $(relrank_mul_rank_top h))
+
+@[simp]
+theorem relrank_top_left : relrank ⊤ A = 1 := relrank_eq_one_of_le le_top
+
+@[simp]
+theorem relfinrank_top_left : relfinrank ⊤ A = 1 := relfinrank_eq_one_of_le le_top
+
+@[simp]
+theorem relrank_top_right : relrank A ⊤ = Module.rank A E := by
+  rw [relrank_eq_rank_of_le (show A ≤ ⊤ from le_top), extendScalars_top,
+    IntermediateField.topEquiv.toLinearEquiv.rank_eq]
+
+@[simp]
+theorem relfinrank_top_right : relfinrank A ⊤ = finrank A E := by
+  simp [relfinrank_eq_toNat_relrank, finrank]
+
+theorem lift_relrank_comap (f : L →+* E) (B : Subfield L) :
+    lift.{v} (relrank (A.comap f) B) = lift.{w} (relrank A (B.map f)) := by
```

Comment:
Thanks for the alternative proof! I'll check them.

> It doesn't seem that you're following the `relindex` API exactly?

That probably because some of the proof for `relindex` does not work in this case. So I changed the order of them.

Reply:
Done.